src/virt/emulator.rs

```diff
@@ -782,18 +791,34 @@ fn has_user_mode(ctx: &VirtContext) -> bool {
     (ctx.csr.misa & misa::U) != 0
 }
 
-/// Return the ID of the next interrupt to be delivered, if any.
-fn get_next_interrupt(mie: usize, mip: usize, mideleg: usize) -> Option<usize> {
-    let ints = mie & mip & !mideleg;
-    if ints == 0 {
-        None
+/// Retrieves the next interrupt by priority similar to the official risc-v specification
+fn find_pending_interrupt_by_priority(ip: usize) -> Option<usize> {
+    if ip & mie::MEIE_FILTER != 0 {
+        Some(MEIE_OFFSET)
+    } else if ip & mie::MSIE_FILTER != 0 {
+        Some(MSIE_OFFSET)
+    } else if ip & mie::MTIE_FILTER != 0 {
+        Some(MTIE_OFFSET)
+    } else if ip & mie::SEIE_FILTER != 0 {
+        Some(SEIE_OFFSET)
+    } else if ip & mie::SSIE_FILTER != 0 {
+        Some(SSIE_OFFSET)
+    } else if ip & mie::STIE_FILTER != 0 {
+        Some(STIE_OFFSET)
     } else {
-        // TODO: use the same priority as hardware.
-        // Currently we serve the less significant bit first.
-        Some(ints.trailing_zeros() as usize)
+        None
     }
 }
 
+/// Return the ID of the next interrupt to be delivered, if any.
+fn get_next_interrupt(mie: usize, mip: usize, mideleg: usize) -> Option<usize> {
+    // First function
+    let ip = mie & mip & !mideleg;
+
+    // Second function (they should be equivalent)
```

Comment:
What does that mean 'thay should be equivalent'?

Reply:
It was for me to debug the formal verification (aka François' brain dump)